applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp

```diff
@@ -79,6 +86,174 @@ class GeoTransportEquationUtilities
+                    Nu(i_dim, index) = nu_container(g_point, i);
+                    index++;
+                }
+            }
+            noalias(aux_density_matrix) = prod(density_matrix, Nu);
+
+            // Adding contribution to Mass matrix
+            const double integration_coefficient_initial_configuration =
+                CalculateIntegrationCoefficientInitialConfiguration(
+                    g_point, rGeom, integration_points, IntegrationMethod, rStressStatePolicy);
+            noalias(mass_matrix_contribution) +=
+                prod(trans(Nu), aux_density_matrix) * integration_coefficient_initial_configuration;
+        }
+
+        // Distribute mass block matrix into the elemental matrix
+        const SizeType element_size = block_element_size + number_P_nodes;
+        Matrix         mass_matrix  = ZeroMatrix(element_size, element_size);
+
+        for (SizeType i = 0; i < number_U_nodes * dimension; ++i) {
+            for (SizeType j = 0; j < number_U_nodes * dimension; ++j) {
+                mass_matrix(i, j) += mass_matrix_contribution(i, j);
+            }
+        }
+        return mass_matrix;
+    }
+
+    static double CalculateFluidPressure(const Vector& rNp, const Vector& rPressureVector)
+    {
+        return inner_prod(rNp, rPressureVector);
+    }
+
+    static Vector GetSolutionVector(SizeType NumberPNodes, const Geometry<Node>& rGeom, const Variable<double>& SolutionVariable)
+    {
+        Vector solution_vector(NumberPNodes);
+        for (SizeType i = 0; i < NumberPNodes; ++i) {
+            solution_vector[i] = rGeom[i].FastGetSolutionStepValue(SolutionVariable);
+        }
+        return solution_vector;
+    }
+
+    static double CalculateSoilDensity(double DegreeOfSaturation, const Properties& rProp)
+    {
+        return (DegreeOfSaturation * rProp[POROSITY] * rProp[DENSITY_WATER]) +
+               (1.0 - rProp[POROSITY]) * rProp[DENSITY_SOLID];
+    }
+
+    template <unsigned int TDim, unsigned int TNumNodes>
+    static Matrix CalculateMassMatrix(unsigned int                              N_DOF,
+                                      const Geometry<Node>&                     rGeom,
+                                      const GeometryData::IntegrationMethod     IntegrationMethod,
+                                      const StressStatePolicy&                  rStressStatePolicy,
+                                      const std::vector<RetentionLaw::Pointer>& rRetentionLawVector,
+                                      const Properties&                         rProp,
+                                      const ProcessInfo&                        rCurrentProcessInfo)
+    {
+        const Geometry<Node>::IntegrationPointsArrayType& integration_points =
+            rGeom.IntegrationPoints(IntegrationMethod);
+        const unsigned int number_G_points = integration_points.size();
+
+        Matrix N_container = rGeom.ShapeFunctionsValues(IntegrationMethod);
```

Comment:
Since the `ShapeFunctionsValues` function returns a const Matrix&, I'd suggest to make this const& to avoid the copy
```suggestion
        const Matrix& N_container = rGeom.ShapeFunctionsValues(IntegrationMethod);
```

Reply:
changed.